Assemble IL in Embedded Resources

delmyers opened the pull request:
<!--
COMMENT BLOCKS WILL NOT BE INCLUDED IN THE PR.
-->

###### Summary <!-- REQUIRED -->
<!-- Quick explanation of the changes. -->
Assemble IL in Embedded Resources

###### Change Log  <!-- REQUIRED -->
<!-- Detail the changes made here. -->
- Example change
1) Copy ilasm as an embedded resource into the InstrEngineTests assembly
2) Add il files to embedded resources
3) Refactor the compiler helpers to extract embedded resources as files
4) Invoke ilasm to assemble the embedded il

Note: I think that ILASM is only available for Windows, so we may need to update our build steps, in the future, to run ILASM for assembly il files, and embed them into the instrumentation engine tests, or bundle them as part of the deployment.

###### Test Methodology
<!-- How was this change validated? i.e. local build, pipeline build etc. -->

Manual testing by executing tests in Visual Studio

Commit:
Assemble IL in Embedded Resources

1) Add ILASM to test project
2) Copy ilasm as an embedded resource into the InstrEngineTests assembly
3) Add il files to embedded resources
4) Refactor the compiler helpers to extract embedded resources as files
5) Invoke ilasm to assemble the embedded il

Note: I think that ILASM is only available for Windows, so we may need
to update our build steps, in the future, to run ILASM for assembly il
files, and embed them into the instrumentation engine tests, or bundle
them as part of the deployment.

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/EmbeddedResourceUtils.cs:
```diff
@@ -0,0 +1,160 @@
+        /// </summary>
+        /// <param name="resourceName">The name of the resource.</param>
+        /// <param name="required">True if an error should occur when the resource is not found.</param>
+        /// <returns>An IEmbenddedResourceFile that has info about the extracted file.</returns>
```
```suggestion
        /// <returns>An IEmbeddedResourceFile that has info about the extracted file.</returns>
```

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/EmbeddedResourceUtils.cs:
```diff
@@ -0,0 +1,160 @@
+            }
+        }
+
+        private static Stream GetEmbeddedResource(string file, bool required = true)
```
nit: maybe use `fileName` or something similar instead of `file` so there's no doubt about whether it's the name/path/contents of the file

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/EmbeddedResourceUtils.cs:
```diff
@@ -0,0 +1,160 @@
+                {
+                    foreach (FileInfo f in resourceDirectory.EnumerateFiles())
+                    {
+                        f.Delete();
```
Something to consider - do we want/need to handle potential `IOExceptions` here, where a deletion might fail?

delmyers replied:
I think we can leave it alone. The tests shouldn't pass if the delete fails. There is something wrong with the infrastructure in that case.

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/InstrEngineTests.csproj:
```diff
@@ -502,9 +502,14 @@
   </ItemGroup>
   <Import Project="$(EnlistmentRoot)\build\Managed.targets" />
+  <ItemGroup>
+    <EmbeddedResource Update="E:\dd\dm\CLRInstrumentationEngine\packages\microsoft.netcore.ilasm\7.0.0-preview.4.22229.4\runtimes\native\ilasm.exe">
```
Do we need to use some sort of relative path here? My guess is this is specific to your machine

delmyers replied:
This was a variable. VS must have replaced it with a full path when I saved the project. This is a bug in VS, I think.

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/TargetAppCompiler.cs:
```diff
@@ -57,61 +55,72 @@ internal class TargetAppCompiler
+            AssembleILTestCode(directoryPath);
+        }
+
+        private static void ComplileCSharpTestCode(string directoryPath)
```
```suggestion
        private static void CompileCSharpTestCode(string directoryPath)
```

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/TargetAppCompiler.cs:
```diff
@@ -38,9 +38,7 @@ internal class TargetAppCompiler
-        private const string EmbeddedResourcesPath = "InstrEngineTests.EmbeddedResources";
-
-        private static string[] TestAppFilePrefixes = new string[] {
+        private static string[] EmbeddedCSharpFiles = new string[] {
```
nit: These are still prefixes (same with `EmbeddedILFiles`), so it might be good to keep it in the name (e.g. `EmbeddedCSharpFilePrefixes`)

kkeirstead commented on src/Tests/InstrEngineTests/InstrEngineTests/TestAppAssembler.cs:
```diff
@@ -0,0 +1,45 @@
+            psi.UseShellExecute = false;
+            string debugSwitch = isDebug ? "/DEBUG=OPT" : string.Empty;
+            string bitSwitch = is64Bit ? "/X64" : "/32BITPREFERRED";
+            string outputType = isExe ? "/EXE" : "/DLL";
```
I'm assuming this behavior is intentional for our specific use-case, but figured I'd point it out - this is the opposite of what you would expect when using `ilasm` - the default is to assume `/EXE` unless `/DLL` is specified, whereas with the default for `isExe` set to false, we automatically default to `/DLL`. Assuming this is the behavior we want by default, perhaps a comment would be useful to call this out.

delmyers replied:
This is by design because it follows the same convension as compiling CSharp. I think with CSharp, we are building exes all the time. We may or may not follow that same decision when we build the assemblies from IL.

Commit:
added links to 1.0.45

willxie-eng commented on src/Tests/InstrEngineTests/InstrEngineTests/EmbeddedResourceUtils.cs:
```diff
@@ -0,0 +1,160 @@
+                    foreach (FileInfo f in resourceDirectory.EnumerateFiles())
+                    {
+                        f.Delete();
+                    }
```
Consider using `resourceDirectory.Delete(recursive: true);`

Commit:
updated link in readme

Commit:
added redirect code as part of alive status codes

Commit:
added ignorepattern for microsoft.com to avoid issues with 'bad' links

Commit:
regex fix

Commit:
Build in the latest VS Preview (#507)

- Add dotnet nuget feeds to nuget.config
  This is needed because the dotnet build tools frequently
  update, and we need access to newer SDKs. It is not
  feasible to continue to upload packages from nuget.org to
  our own feed.
- Fix some issues with code analysis.
  The VS code analysis tools are moving away from ruleset
  suppressions, and toward suppressions in code and in .editorconfig
  files. We add our global suppressions to the .editorconfig,
  and make some more targeted suppressions inside the affected
  code, or fix the error found by code analysis. The ruleset is
  still here for compatibility with older VS.
- Change the <Assembly>_<arch> projects to be <Assembly>.<arch>
  instead. This is due to a bug in codeanalysis, which is no
  longer suppressing the rule that disallows underscores in
  assembly names. Since this is blocking our build, and the
  rule only applies to our test binaries, this PR just works
  around the bug.

Commit:
escape dot after www

Commit:
Merge branch 'main' into dev/gophoon/release1.0.45

Commit:
removed changes to trademarks section

Commit:
Merge branch 'dev/gophoon/release1.0.45' of https://github.com/microsoft/CLRInstrumentationEngine into dev/gophoon/release1.0.45

Commit:
Merge pull request #506 from microsoft/dev/gophoon/release1.0.45

added links to 1.0.45

delmyers commented:
closing this pull request in favor of the one in the main CLRIE repo: https://github.com/microsoft/CLRInstrumentationEngine/pull/508
